### ISSUE TITLE
Warning on `setup.py` being deprecated

### ISSUE THREAD

**Issue by mtazzari** [OasisLMF/OasisLMF#1158]
## Issue Description
When installing `oasislmf` with the verbose flag:
```bash
cd OasisLMF/
pip install -v . 
```
I have noticed the following warning:
```bash
[...]/envs/oasis/lib/python3.10/site-packages/setuptools/command/install.py:34: 
SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
```

## Version / Environment information
Linux 5.15.0-53-generic Ubuntu
OasisLMF 1.27.0rc3

**Comment by sr-murthy** [comment]
Hi @sambles @mtazzari Hello :)

May I suggest switching to [`pyproject.toml`](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/) and a new style packaging/build manager such as [PDM](https://pdm.fming.dev/latest/)? PDM supports an [import command](https://pdm.fming.dev/latest/reference/cli/#import) that makes easy to generate an initial `pyproject.toml`, which you can then customise.

For some time now, `setup.py` based configuration and builds have been [discouraged](https://pip.pypa.io/en/stable/reference/build-system/setup-py/). It is possible to completely replace `setuptools`, `wheel`, and even `pip`, with tools like PDM.

**Comment by sr-murthy** [comment]
[@sambles](url) The continued presence of `setup.py` in this project, without an appropriate [project TOML](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/), poses a number of risks for this project, both for end users as well as maintainers, developers, and contributors:

* The complete failure of editable installs (via `pip install -e -> python setup.py develop`) by August 2025, because of a [scheduled pip change](https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#setup-py-deprecated).
* The potential failure of standard (non-editable) installs, via `pip install -> python setup.py install` - deprecation warnings regularly appear for package installations. There is a simple solution which I would outline below.
* The potential failure of wheel builds via `python setup.py bdist_wheel` - similarly, errors and deprecation warnings regularly appear during package installations, and again there is a simple solution for this, which will also deal with point above.

I'll try and describe some of the relevant and important details first, then outline some suggestions.

1. Firstly, in relation to the editable installation issue, note that it's not `pip install -e` that is being deprecated, but the fallback behaviour of `pip install -e` triggering `python setup.py develop` for projects, such as MDK, which contain only a `setup.py` without a [project TOML defining setuptools as the build backend](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#should-pyproject-toml-be-added).  In summary, after August 2025 `pip install -e oasislmf` will fail, as long as the project does not contain a project TOML defining a build backend. This is a breaking change that may impact an unknown number of users and/or contributors, depending on how people are using the project and package. It's definitely something to be aware of.

You could argue that the issue of editable installations is not such a huge problem for this project.

2. But there are also other, more [general risks to using `setup.py`](https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#setup-py-deprecated), starting with standard (non-editable) installations of this package, via `python setup.py install` - at the moment, this is exactly how standard installations happen (via `pip install oasislmf`), and you see deprecation warnings such as:
```bash
/home/srm/.pyenv/versions/OasisEnv_1_28_2/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:79: SetuptoolsDeprecationWarning: setup.py install is deprecated.
      !!
      
              ********************************************************************************
              Please avoid running ``setup.py`` directly.
              Instead, use pypa/build, pypa/installer or other
              standards-based tools.
      
              See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
              ********************************************************************************
```
For more information on why `setup.py install` is deprecated see [this](https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#is-setup-py-deprecated). My own view is that if the fallback behaviour of `python setup.py develop` can be removed for projects without a project TOML, so can support for `python setup.py install`. I wouldn't be surprised if `pip install` will stop supporting `python setup.py install` from 2026, in which case `pip install oasislmf` will also fail. As with the [editable installations issue](https://github.com/pypa/pip/issues/11457) the `pip` maintainers will make the necessary changes to reduce the maintenance overload on supporting legacy features of `setuptools`, which is becoming considerable.

3. The other risk is the potential failure of wheel builds, via `python setup.py bdist_wheel`, which now generates the following series of horrible errors and deprecation warnings:
```bash
  Building wheel for oasislmf (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [1596 lines of output]
      /home/srm/.pyenv/versions/OasisEnv_1_28_2/lib/python3.10/site-packages/setuptools/dist.py:493: SetuptoolsDeprecationWarning: Invalid dash-separated options
      !!
      
              ********************************************************************************
              Usage of dash-separated 'description-file' will not be supported in future
              versions. Please use the underscore name 'description_file' instead.
      
              By 2025-Mar-03, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.
      
              See https://setuptools.pypa.io/en/latest/userguide/declarative_config.html for details.
              ********************************************************************************
      
      !!
        opt = self.warn_dash_deprecation(opt, section)
      running bdist_wheel
      The [wheel] section is deprecated. Use [bdist_wheel] instead.
      /tmp/pip-install-qs6pv_qw/oasislmf_ccc74468767a4435bcfca3962cb4d9b0/setup.py:291: SetuptoolsDeprecationWarning: bdist_wheel.universal is deprecated
      !!
      
              ********************************************************************************
              With Python 2.7 end-of-life, support for building universal wheels
              (i.e., wheels that support both Python 2 and Python 3)
              is being obviated.
              Please discontinue using this option, or if you still need it,
              file an issue with pypa/setuptools describing your use case.
      
              By 2025-Aug-30, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.
              ********************************************************************************
      
      !!
...
```

An alternative and better approach to wheel builds is to use a dedicated build frontend, such as [`build`](https://build.pypa.io/en/stable/), which is developed by PyPA, or any number of alternatives including [PDM](https://pdm-project.org/), [Hatch](https://hatch.pypa.io/1.13/), [Poetry](https://python-poetry.org/) etc. Of these I strongly recommend PDM, as it is more fully PEP compliant than any other available alternative. Note that all of these tools require the use of the project TOML, which has now become a best practice for Python packaging.

There are also other setuptools-related errors and deprecation warnings that I have seen in MDK installations, including this one about `oasislmf._data`:
```bash
      /home/srm/.pyenv/versions/OasisEnv_1_28_2/lib/python3.10/site-packages/setuptools/command/build_py.py:212: _Warning: Package 'oasislmf._data' is absent from the `packages` configuration.
      !!
      
              ********************************************************************************
              ############################
              # Package would be ignored #
              ############################
              Python recognizes 'oasislmf._data' as an importable package[^1],
              but it is absent from setuptools' `packages` configuration.
      
              This leads to an ambiguous overall configuration. If you want to distribute this
              package, please make sure that 'oasislmf._data' is explicitly added
              to the `packages` configuration field.
      
              Alternatively, you can also rely on setuptools' discovery methods
              (for example by using `find_namespace_packages(...)`/`find_namespace:`
              instead of `find_packages(...)`/`find:`).
      
              You can read more about "package discovery" on setuptools documentation page:
      
              - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html
      
              If you don't want 'oasislmf._data' to be distributed and are
              already explicitly excluding 'oasislmf._data' via
              `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,
              you can try to use `exclude_package_data`, or `include-package-data=False` in
              combination with a more fine grained `package-data` configuration.
      
              You can read more about "package data files" on setuptools documentation page:
      
              - https://setuptools.pypa.io/en/latest/userguide/datafiles.html
      
      
              [^1]: For Python, any directory (with suitable naming) can be imported,
                    even if it does not contain any `.py` files.
                    On the other hand, currently there is no concept of package data
                    directory, all directories are treated like packages.
              ********************************************************************************
```

A further general point on these deprecation warnings is that too many of them create not just messy and overlong pipeline logs, but create the risk of a perceived decrease in package quality and reliability.

Packaging is a longstanding headache for many Python projects, but there are a few simple steps that can be taken to control the risks of using legacy tools such as `setup.py`. I would recommend starting with the following documentation on understanding the scope of the risks associated with using `setup.py`:

* https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#is-setup-py-deprecated

and this set of recommendations for migrating to a better packaging strategy:

* https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#should-pyproject-toml-be-added

The specific initial recommendation I would make is to add a `pyproject.toml` that includes basic package metadata as well as a definition of the build system using `setuptools`, e.g. the following lines of TOML:
```toml
[build-system]
# XXX: If your project needs other packages to build properly, add them to this list.
requires = ["setuptools >= 64"]
build-backend = "setuptools.build_meta"
```
For more information on writing a project TOML you may consult [this official guide](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/). PyPA also provide a [helpful glossary](https://packaging.python.org/en/latest/glossary/), which is very useful on understanding the now very complex and evolved terminology of Python packaging.

I would also be happy to submit an evaluation PR containing the relevant changes which you could review, including the testing of built artifacts such as source distributions and wheels.

**Comment by ichard26** [comment]
@sr-murthy Technically, it's not guaranteed that once the legacy mechanism is removed, `pyproject.toml` is required. As explained in the issue description, pip will fallback to a default `pyproject.toml` which includes setuptools. There's a solid chance that if you do nothing, things will continue to work. You can try today with `--use-pep517` to force the modern mechanism (w/ the fallback `pyproject.toml`).

_\<rant> I wrote that issue description, but seems like it was totally useless. I get the feeling that no one actually understands what's going on, and maybe I should've taken a whole another approach to the issue description, assuming our users know absolutely nothing about packaging. OTOH, we really want people to move to `pyproject.toml`. \</rant>_

**Comment by sr-murthy** [comment]
@ichard26 Yes, more clarity would be better, yet we have the official PyPA recommendation to stop using `setup.py` commands, including `develop` - they are all deprecated, right? My point is broader: should you take chances on make assumptions? Terms such as "solid chance" are not reassuring, and you don't want to take chances if you've got production pipelines depending on a package with these issues.

**Comment by ichard26** [comment]
> PyPA recommendation to stop using setup.py commands, including develop - they are all deprecated, right

Yes. They are. That's a big part of why pip wants to stop calling them internally.

> should you take chances on making assumptions?

It's up to every individual project. We can't make that decision for you. The issue is written with the "best solution" for most which is to indeed add a `pyproject.toml`. 

> more clarity would be better

Yea, that's fair. I have no more interest in fixing this as it seems fruitless. I don't participate in PyPA level discussions because I don't have the energy to deal with that. I am a pip maintainer, and that's all. I just don't want people to panic because that's what makes everyone angry and hate us for "breaking everything". Some panic is justified, but IMO not "it's going to 100% break for sure".

Anyway. I'm being an intrusion to a third-party project. I'll go now.

**Comment by sr-murthy** [comment]
Np @ichard26 not an intrusion, your thoughts are welcome. I think pip is indispensable, so best wishes for your work on maintaining it.

**Comment by sambles** [comment]
Hi @sr-murthy 

Thanks for the feedback,  a packaging refactor is needed and planned, but we are first aiming to drop the C++ component in favour of python/numba rewritten equivalents.  

I think updating once all those new tools are in place makes the most sense, since we can strip out all of the C++ compilation and binary inclusion all together.  (A contingency build option might be required for modellers still using ktools. ~ not sure yet)

Progress is being made on that front, and we expect to have things ready later this year.

**Comment by sr-murthy2** [comment]
Sounds good. If you need any further feedback, evaluation etc., esp. in relation to the project TOML-side of things, I  would be happy to contribute. I've switched to using project TOMLs + PDM in my projects, and they work very well.

**Comment by sambles** [comment]
fixed
